### DISCUSSION TITLE
Naming discuss

### DISCUSSION THREAD

**Discussion by panno8M** [godot-nim/.github#11] [General]
@gushromp @insomniacUNDERSCORElemon 

I don't think the current repository name correctly reflects the architecture, so I would like to rename it again before registering it with nim-lang/packages.

To this end, I will explain the current repository structure and its intent. Currently there are two essential repositories (packages): gdext-nim (gdext.nimble) and gdextgen (gdextgen.nimble).

The two are separated by whether or not they depend on a specific version of Godot.
gdextgen has branches for versions corresponding to Godot, such as 4.3 and 4.2, and the user specifies the desired version and installs it.
In contrast, gdext-nim consists of a core layer that abstracts version differences and a surface layer that wraps the core layer and gdextgen and takes usability into consideration.

Next, use cases after registration to nim-lang/packages will be described. Both gdext-nim and gdextgen will be registered to packages.
Have users install gdextgen as follows:

```console
$ nimble install gdextgen
```
```console
$ nimble install gdextgen@#4.3
```

gdextgen.nimble requires gdext-nim as follows, so you can set up gdext by simply installing gdextgen.

```nim
# gdextgen.nimble
requires "gdext >= 1.0.0"
```

The above is the information to be shared for this discussion. Here is the main issue.

1. I am considering renaming the following four. Please give me your suggestions.

   * repository godot-nim/gdext-nim
   * package gdext (hosted on godot-nim/gdext-nim)
   * repository godot-nim/gdextgen
   * package gdextgen (hosted on godot-nim/gdextgen)
   
   I have no intention of changing the project name (gdext) and would like to include gdext in the above four names.
  
2. If you have any suggestions for improvements to the package architecture, please let me know.

Thanks,

**Comment by insomniacUNDERSCORElemon** [comment]
Maybe I'm missing something, but the names seem fine to me... gdext is in all 4 of those lines already.

The only thing lacking it would be godot-nim, which on top of it being the org name I don't think the redundancy is needed there.

**Reply by panno8M** [reply]
It has been a while. Thanks for the reply.

I don't think it is that much of a problem for gdext-nim, but I think there is room for improvement for gdextgen.
From the user's point of view, it is not important whether this package is generated or not, and it would be more meaningful to know that it is a layer that depends on the version of godot.
I would like to give it a name that conveys this kind of context, but my English vocabulary is not very good, so I was hoping to get some ideas.

**Reply by insomniacUNDERSCORElemon** [reply]
Alright, that makes more sense. I am not sure that _generated_ is bad here... I agree this might not tell a user much, but on-the-other-hand anything generated somewhat implies that it will be version specific.

Would (API) mappings be an applicable term here? Something like `gdext-mapgen`, or `gdext-apigen` (if not just `gdext-mappings`)? Assuming those even apply, I am not sure how to make version specificity more obvious than that (particularly in the case that it is the piece that needs to be updated for new versions), but then again this sort of piece has always seemed to me like something that was more of an under-the-hood thing that users likely don't need to think about.

I think adapter (or shim?) could be potentially used too, though even something that makes sense when thinking about it might not be intuitive to most users.

**Reply by panno8M** [reply]
umm, the keyword “gen” is meaningful to developers, but I think it is noise to users.
To set up gdext, I assume the following commands:

```console
; Package for the latest gdext-nim, compatible with the latest godot stable
$ nimble install gdextgen

; Package for the latest gdext-nim, compatible with godot 4.3 stable
$ nimble install gdextgen@#4.3

; Package for gdext-nim 1.0.0, compatible with godot 4.3 stable
$ nimble install gdextgen@#4.3-1.0.0
```

Specifiers beginning with `#` represent branches, tags, commit hashes, etc. on the repository.

I suspect that when presented with this snippet, users will not understand what this package is all about; could it be not?

I want to name it which users can understand at a glance that “this package is a component to enable gdext-nim to work with godot 4.3”. I'm not sure that the word “mapping” meets this requirement.

**Reply by insomniacUNDERSCORElemon** [reply]
I didn't think about this being about nimble, I'm not sure this is the best way to go for the naming/structure.

Maybe you could copy choosenim a bit? Not handling the versioning with the initial package install, also you could add detecting the version of the user's Godot install. I assume allowing the user to specify their target would not be too difficult, but again I think just rolling with stable right now is a pretty safe/expected workflow until far-future Godot/Gdext(-nim) changes actually create a usage split (also I assume that a user ignoring Godot updates would have the desired effect for gdext as well).

**Reply by panno8M** [reply]
> Maybe you could copy choosenim a bit? Not handling the versioning with the initial package install, also you could add detecting the version of the user's Godot install. I assume allowing the user to specify their target would not be too difficult, but again I think just rolling with stable right now is a pretty safe/expected workflow until far-future Godot/Gdext(-nim) changes actually create a usage split (also I assume that a user ignoring Godot updates would have the desired effect for gdext as well).

For this, I plan to provide a `switch` (or `choose`) command in gdextwiz. The process is simply to uninstall the currently installed gdextgen and reinstall the specified or detected gdextgen.
If nothing is specified, "rolling with stable" will done, so I don't really understand the concerns you point out.

I rather would have liked to receive a candidate for the name...

**Reply by insomniacUNDERSCORElemon** [reply]
> I rather would have liked to receive a candidate for the name...

My point was that I don't think the naming is super important, particularly when it likely shouldn't be relevant during initial install. That, and you could potentially even go with a name like choosegdext which should make the versioning aspect very obvious to most users due to choosenim being part of the Nim ecosystem.

**Reply by panno8M** [reply]
Ah I see. I will refer to your suggestion. Thank you.
